modules/concept-docs/pages/n1ql-query.adoc

```diff
@@ -14,6 +14,47 @@ Below, we fill in some of the gaps between reference and rolling-up-your-sleeves
 with discussion of a few areas of the Query service where more background knowledge will help you to better program your application.
 
 
+== Prepared Statements for Query Optimization
+
+When a N1QL query string is sent to the server, the server will inspect the string and parse it, planning which indexes to query.
+Once this is done, it generates a _query plan_.
+The computation for the plan adds some additional processing time and overhead for the query.
+
+Often-used queries can be _prepared_ so that its _plan_ is generated only once.
+Subsequent queries using the same query string will use the pre-generated plan instead, saving on the overhead and processing of the plan each time.
+
+You can indicate to the SDK that a given query should be optimized in the above fashion.
+When an SDK sees that a query should be optimized, it will internally prepare the statement and store the plan in an internal cache.
+When issuing the query again, the SDK will check to see if a plan exists in its cache, and will send the plan to the server.
+
+To indicate that an SDK should optimize a query, the `Adhoc` parameter should be set to false.
+When a query is not ad hoc, the SDK will fetch the plan (if it does not already have it).
+Do not turn off the `Adhoc` flag for each query since only a finite number of query plans (currently 5000) can be stored in the SDK.
```

Comment:
I'm pretty sure we don't actually have any hard limit on plans.

Reply:
Thanks @chvck - will dig into this.

Reply:
`private static final int QUERY_CACHE_SIZE = 5000;`
in Java SDK 3.0,
but I'm changing the text in any case  :+1: